### ISSUE TITLE
SC2295 wiki explanation maybe not clear enough?

### ISSUE THREAD

**Issue by oliv3r** [koalaman/shellcheck#2385]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2295
- My shellcheck version (`shellcheck --version` or "online"): online
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [ ] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [ ] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/sh
foo1="${bar:-'default value'}"
foo2="${bar:-"default value"}"
```

On the SC2295 wiki page, it is stated that "Expansions inside ${..} need to be quoted separately, otherwise they will match as a pattern." But I'm not sure that is accurate. If we look at the examples `foo1` where we single quote a string, then `foo1` will actually contain `'default value'`. For `foo2` I fear it's more subtle, and afraid it also doesn't do what one would expect. I believe in that case, we end do not end up with `"default value"` but rather the initial opening quote gets closed. to use square brackets as syntactic example, where write `foo2=[${bar:-[default value]}]` we end up with `foo2=[${bar:-]default value[}]`, e.g. we close the first quote instead of opening a sub-quote (if that makes sense). Which is most possibly not what we would want. I think the result is not so bad, we just end up with `foo2=default value` e.g. an unquoted string, and I think in shell, this doesn't even matter if it would contain spaces.

For the `#` `##` `%` `%%` cases the SC2295 is correct.

P.S. I'm opening this ticket to get clarification, so I can learn what to (if) modify on the wiki page.

**Comment by brother** [comment]
WFM
```
brother ~$ dash
$ foo1="${bar:-'default value'}"
$ foo2="${bar:-"default value"}"
$ echo ".$foo1."
.'default value'.
$ echo ".$foo2."
.default value.
$ foo3="${bar:-default value}"
$ echo ".$foo3."        
.default value.
$ 
```

The case of SC2295 is about expansions and this example doesn't really use them but the quoting seems to work as intended. No?

**Comment by oliv3r** [comment]
SC2295 itself works fine; I'm talking mostly about the wikipage. Right now, the wikipage treats it as a general thing, but I don't see how the wikipage is applicable to anything that are not `#` `##` `%` or `%%` expansions no? I mean, would `foo1="${bar:-*[something]}" ` be ever anything other then a string?

What I am really after, is to make sure the wiki page regarding SC2295 is 'more' clear; as I still don't get all the cases where you want/want not expansions quoted.

As for the example above, The discussion probably is just academic :p but if you have a generic 'codestyle rule' that sates, all strings should be quoted, and due to SC2295 all expansions should be quoted as well by default (to generalize); well you can't single-quote strings because of the expansion (foo1), and double quoting them is also 'wrong' (foo2) as yes, it works, but I think for the wrong reasons. E.g. you are putting an unquoted variable into `foo2` (again just academic at this point, as the result is the same).

**Comment by brother** [comment]
Ok got it. I am not sure it's possible to wirte a clear description to cover it all.

Demanding a specific quote style is pretty obvious a bad idea, you have all the proofs needed for that discussion if you want.

**Comment by oliv3r** [comment]
Fair enough; but still, the wiki page is off right now, capturing it better is hard of course. So from my understanding, as I said, the only cases that need quoting, where the shell will expand them away, are the `#`, `##`, `%` and `%%` cases, as everything else is considered a string already anyway.

**Comment by koalaman** [comment]
I'm not sure I follow. Are you saying that SC2295's suggestion to quote expansions in parameter substitutions does not apply to default values like `foo1="${bar:-'default value'}"` (which as far as I can see also does not trigger SC2295)? 

It does qualify with "When using expansions in a parameter expansion prefix/suffix expression [..]" and any implication that it should apply to `:-` as well is not intentional.

**Comment by oliv3r** [comment]
@koalaman so shellcheck does the right thing :) However the wiki page lacks to clearly explain the why/whats. I had a bit of trouble understanding really what/why was going on, and so looked at the wiki page to make sure to define codestyle rules to follow my self. My first 'knee-jerk' reaction was. Ah, we should quote everything.

e.g.
`foo1="${bar:-'default value'}"` because it's a string and `foo2="${bar2:-"default $foo1:-} value"}"` because it contains a variable as part of its string. The `foo2` case worked fine, but for the wrong reasons imo, the foo1 case now printed the additional quotes.

The SC2295 wiki page however makes you think all _should_ be quoted, as it just mentions variable expansion. I'm arguing, the wiki should reflect, that this is only needed on expansions that involve `#`, `##`, `%` or `%%`.

I didn't want to edit the wiki page without some discussion here, as I (initially?) lacked full understanding of things.
